File: claude-code-python/src/claude_code/utils/hash.py
"""Hash utilities. Ported from utils/hash.ts"""
from __future__ import annotations
import hashlib


def djb2_hash(s: str) -> int:
    """djb2 string hash — fast non-cryptographic hash returning a signed 32-bit int.

    Deterministic across runtimes (unlike platform-specific hash seeds).
    Use as a fallback when a stable, on-disk-safe hash is needed
    (e.g., cache directory names that must survive runtime upgrades).

    Ported from utils/hash.ts: djb2Hash.
    """
    h = 0
    for ch in s:
        h = ((h << 5) - h + ord(ch)) & 0xFFFFFFFF
    # Convert to signed 32-bit int (mirror the TypeScript ``| 0`` behaviour)
    if h >= 0x80000000:
        h -= 0x100000000
    return h


def hash_content(content: str) -> str:
    """Hash arbitrary content for change detection.

    Uses SHA-256 (fast enough for diff detection; not cryptographically
    sensitive). Mirrors the Node.js path in utils/hash.ts: hashContent.
    """
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


def hash_pair(a: str, b: str) -> str:
    """Hash two strings without allocating a concatenated temporary string.

    Uses incremental SHA-256 update with a NUL separator to distinguish
    ``("ts", "code")`` from ``("tsc", "ode")``.  Mirrors the Node.js path in
    utils/hash.ts: hashPair.
    """
    h = hashlib.sha256()
    h.update(a.encode("utf-8"))
    h.update(b"\x00")
    h.update(b.encode("utf-8"))
    return h.hexdigest()

File: claude-code-rust/crates/claude-utils/src/hash.rs
// 原始 TS: utils/hash.ts
//! Non-cryptographic and cryptographic hash utilities.
//!
//! Uses `seahash` for fast non-crypto hashing (equivalent to djb2/Bun.hash)
//! and `sha2` for crypto-quality hashing.

use std::hash::{Hash, Hasher};

/// djb2 string hash — fast non-cryptographic hash returning a signed 32-bit integer.
/// Deterministic across runtimes. Use when you need stable on-disk output.
pub fn djb2_hash(s: &str) -> i32 {
    let mut hash: i32 = 0;
    for c in s.bytes() {
        hash = hash.wrapping_shl(5).wrapping_sub(hash).wrapping_add(c as i32);
    }
    hash
}

/// Hash arbitrary content for change detection.
/// Uses a fast non-crypto hash (seahash). Not suitable for security purposes.
pub fn hash_content(content: &str) -> u64 {
    seahash::hash(content.as_bytes())
}

/// Hash arbitrary content to a hex string using SHA-256 (crypto-quality).
pub fn hash_content_sha256(content: &str) -> String {
    use sha2::{Digest, Sha256};
    let mut hasher = Sha256::new();
    hasher.update(content.as_bytes());
    format!("{:x}", hasher.finalize())
}

/// Hash two strings without allocating a concatenated temp string.
/// Uses seahash seeding to naturally disambiguate ("ts","code") vs ("tsc","ode").
pub fn hash_pair(a: &str, b: &str) -> u64 {
    // Seed the second hash with the first hash value
    let seed_a = seahash::hash(a.as_bytes());
    // XOR with a separator to distinguish positions
    let seed_b = seahash::hash_seeded(b.as_bytes(), seed_a, 0, 0, 0);
    seed_b
}

/// Hash a pair to a hex string using SHA-256 (stable, crypto-quality).
pub fn hash_pair_sha256(a: &str, b: &str) -> String {
    use sha2::{Digest, Sha256};
    let mut hasher = Sha256::new();
    hasher.update(a.as_bytes());
    hasher.update(b"\0"); // separator
    hasher.update(b.as_bytes());
    format!("{:x}", hasher.finalize())
}

/// Hash a generic value using Rust's standard `Hash` trait + seahash.
pub fn hash_value<T: Hash>(value: &T) -> u64 {
    let mut hasher = seahash::SeaHasher::new();
    value.hash(&mut hasher);
    hasher.finish()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_djb2_hash_deterministic() {
        assert_eq!(djb2_hash("hello"), djb2_hash("hello"));
        assert_ne!(djb2_hash("hello"), djb2_hash("world"));
    }

    #[test]
    fn test_hash_content_deterministic() {
        assert_eq!(hash_content("hello"), hash_content("hello"));
        assert_ne!(hash_content("hello"), hash_content("world"));
    }

    #[test]
    fn test_hash_pair_order_matters() {
        // ("ts","code") should differ from ("tsc","ode")
        assert_ne!(hash_pair("ts", "code"), hash_pair("tsc", "ode"));
        // Symmetric: same order = same hash
        assert_eq!(hash_pair("a", "b"), hash_pair("a", "b"));
    }

    #[test]
    fn test_hash_content_sha256() {
        let h = hash_content_sha256("hello");
        assert_eq!(h.len(), 64); // hex SHA-256
        assert_eq!(h, hash_content_sha256("hello"));
    }
}
